### DISCUSSION TITLE
Toby's Random Thought #2 - first-class types

### DISCUSSION THREAD

**Discussion by tobyink** [Perl-Oshun/oshun#6] [General]
Under this proposal, types don't seem to be first class. There's no way to do, for example:

```perl
my %types = (
  numbers    => NUM,
  integers   => INT,
);

while ( my ( $name, $type ) = each %types ) {
  no strict 'refs';
  *{"add_$name"} = sub ( $x :of($type), $y :of($type) ) { $x + $y };
}

say add_integers( 40, 2 );
```

Having types be first class allows them to be passed around to things like database ORMs or HTML form validators. For example, it might be useful for an ORM to know that if a field is an `INT`, it should use a particular SQL data type, and if you're generating an HTML form, to use `<input type=number>` to get input for that field.

**Comment by Ovid** [comment]
Part of the issue with what you're suggesting is due to how attributes work. The `$type` in `:of($type)` is literally the string `$type` and it's a compile-time thing, not runtime. I've advocated on P5P to rethink how attributes work, but have gotten no traction there.

**Comment by leonerd** [comment]
Yup. Types need to be first-class.
